### PR TITLE
feat: add libp2p built-in metrics to /metrics endpoint

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- add libp2p built-in metrics to /metrics endpoint

The new section looks like:
```
# SECTION libp2p
# HELP libp2p_gossipsub_messages Number of messages received.
# TYPE libp2p_gossipsub_messages counter
libp2p_gossipsub_messages_total 24
# HELP libp2p_identify_protocols Number of connected nodes supporting a specific protocol, with "unrecognized" for each peer supporting one or more unrecognized protocols.
# TYPE libp2p_identify_protocols gauge
libp2p_identify_protocols{protocol="unrecognized"} 45
libp2p_identify_protocols{protocol="/ipfs/id/1.0.0"} 45
libp2p_identify_protocols{protocol="/ipfs/id/push/1.0.0"} 45
libp2p_identify_protocols{protocol="/ipfs/ping/1.0.0"} 45
# HELP libp2p_identify_errors Number of errors while attempting to identify the remote.
# TYPE libp2p_identify_errors counter
libp2p_identify_errors_total 0
# HELP libp2p_identify_pushed Number of times identification information of the local node has been actively pushed to a peer..
# TYPE libp2p_identify_pushed counter
libp2p_identify_pushed_total 0
# HELP libp2p_identify_received Number of times identification information has been received from a peer.
# TYPE libp2p_identify_received counter
libp2p_identify_received_total 80
# HELP libp2p_identify_received_info_listen_addrs Number of listen addresses for remote peer received in identification information.
# TYPE libp2p_identify_received_info_listen_addrs histogram
libp2p_identify_received_info_listen_addrs_sum 120.0
libp2p_identify_received_info_listen_addrs_count 80
libp2p_identify_received_info_listen_addrs_bucket{le="0.0"} 0
libp2p_identify_received_info_listen_addrs_bucket{le="1.0"} 42
libp2p_identify_received_info_listen_addrs_bucket{le="2.0"} 78
libp2p_identify_received_info_listen_addrs_bucket{le="4.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="8.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="16.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="32.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="64.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="128.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="256.0"} 80
libp2p_identify_received_info_listen_addrs_bucket{le="+Inf"} 80
# HELP libp2p_identify_received_info_protocols Number of protocols supported by the remote peer received in identification information.
# TYPE libp2p_identify_received_info_protocols histogram
libp2p_identify_received_info_protocols_sum 1564.0
libp2p_identify_received_info_protocols_count 80
libp2p_identify_received_info_protocols_bucket{le="0.0"} 0
libp2p_identify_received_info_protocols_bucket{le="1.0"} 0
libp2p_identify_received_info_protocols_bucket{le="2.0"} 0
libp2p_identify_received_info_protocols_bucket{le="4.0"} 0
libp2p_identify_received_info_protocols_bucket{le="8.0"} 0
libp2p_identify_received_info_protocols_bucket{le="16.0"} 0
libp2p_identify_received_info_protocols_bucket{le="32.0"} 80
libp2p_identify_received_info_protocols_bucket{le="64.0"} 80
libp2p_identify_received_info_protocols_bucket{le="128.0"} 80
libp2p_identify_received_info_protocols_bucket{le="256.0"} 80
libp2p_identify_received_info_protocols_bucket{le="+Inf"} 80
# HELP libp2p_identify_sent Number of times identification information of the local node has been sent to a peer in response to an identification request.
# TYPE libp2p_identify_sent counter
libp2p_identify_sent_total 100
# HELP libp2p_kad_query_result_get_record_ok Number of records returned by a successful Kademlia get record query..
# TYPE libp2p_kad_query_result_get_record_ok histogram
libp2p_kad_query_result_get_record_ok_sum 0.0
libp2p_kad_query_result_get_record_ok_count 0
libp2p_kad_query_result_get_record_ok_bucket{le="1.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="2.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="4.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="8.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="16.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="32.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="64.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="128.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="256.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="512.0"} 0
libp2p_kad_query_result_get_record_ok_bucket{le="+Inf"} 0
# HELP libp2p_kad_query_result_get_record_error Number of failed Kademlia get record queries..
# TYPE libp2p_kad_query_result_get_record_error counter
# HELP libp2p_kad_query_result_get_closest_peers_ok Number of closest peers returned by a successful Kademlia get closest peers query..
# TYPE libp2p_kad_query_result_get_closest_peers_ok histogram
libp2p_kad_query_result_get_closest_peers_ok_sum 0.0
libp2p_kad_query_result_get_closest_peers_ok_count 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="1.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="2.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="4.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="8.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="16.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="32.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="64.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="128.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="256.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="512.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="+Inf"} 0
# HELP libp2p_kad_query_result_get_closest_peers_error Number of failed Kademlia get closest peers queries..
# TYPE libp2p_kad_query_result_get_closest_peers_error counter
# HELP libp2p_kad_query_result_get_providers_ok Number of providers returned by a successful Kademlia get providers query..
# TYPE libp2p_kad_query_result_get_providers_ok histogram
libp2p_kad_query_result_get_providers_ok_sum 0.0
libp2p_kad_query_result_get_providers_ok_count 0
libp2p_kad_query_result_get_providers_ok_bucket{le="1.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="2.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="4.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="8.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="16.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="32.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="64.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="128.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="256.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="512.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="+Inf"} 0
# HELP libp2p_kad_query_result_get_providers_error Number of failed Kademlia get providers queries..
# TYPE libp2p_kad_query_result_get_providers_error counter
# HELP libp2p_kad_query_result_num_requests Number of requests started for a Kademlia query..
# TYPE libp2p_kad_query_result_num_requests histogram
# HELP libp2p_kad_query_result_num_success Number of successful requests of a Kademlia query..
# TYPE libp2p_kad_query_result_num_success histogram
# HELP libp2p_kad_query_result_num_failure Number of failed requests of a Kademlia query..
# TYPE libp2p_kad_query_result_num_failure histogram
# HELP libp2p_kad_query_result_duration_seconds Duration of a Kademlia query..
# TYPE libp2p_kad_query_result_duration_seconds histogram
# UNIT libp2p_kad_query_result_duration_seconds seconds
# HELP libp2p_kad_routing_updated Number of peers added, updated or evicted to, in or from a specific kbucket in the routing table.
# TYPE libp2p_kad_routing_updated counter
# HELP libp2p_kad_inbound_requests Number of inbound requests.
# TYPE libp2p_kad_inbound_requests counter
# HELP libp2p_ping_rtt_seconds Round-trip time sending a 'ping' and receiving a 'pong'.
# TYPE libp2p_ping_rtt_seconds histogram
# UNIT libp2p_ping_rtt_seconds seconds
libp2p_ping_rtt_seconds_sum 69.53495317999998
libp2p_ping_rtt_seconds_count 193
libp2p_ping_rtt_seconds_bucket{le="0.001"} 0
libp2p_ping_rtt_seconds_bucket{le="0.002"} 0
libp2p_ping_rtt_seconds_bucket{le="0.004"} 0
libp2p_ping_rtt_seconds_bucket{le="0.008"} 0
libp2p_ping_rtt_seconds_bucket{le="0.016"} 0
libp2p_ping_rtt_seconds_bucket{le="0.032"} 0
libp2p_ping_rtt_seconds_bucket{le="0.064"} 0
libp2p_ping_rtt_seconds_bucket{le="0.128"} 0
libp2p_ping_rtt_seconds_bucket{le="0.256"} 48
libp2p_ping_rtt_seconds_bucket{le="0.512"} 176
libp2p_ping_rtt_seconds_bucket{le="1.024"} 191
libp2p_ping_rtt_seconds_bucket{le="2.048"} 193
libp2p_ping_rtt_seconds_bucket{le="+Inf"} 193
# HELP libp2p_ping_failure Failure while sending a 'ping' or receiving a 'pong'.
# TYPE libp2p_ping_failure counter
# HELP libp2p_ping_pong_received Number of 'pong's received.
# TYPE libp2p_ping_pong_received counter
libp2p_ping_pong_received_total 0
# HELP libp2p_swarm_connections_incoming Number of incoming connections.
# TYPE libp2p_swarm_connections_incoming counter
libp2p_swarm_connections_incoming_total 0
# HELP libp2p_swarm_connections_incoming_error Number of incoming connection errors.
# TYPE libp2p_swarm_connections_incoming_error counter
# HELP libp2p_swarm_new_listen_addr Number of new listen addresses.
# TYPE libp2p_swarm_new_listen_addr counter
libp2p_swarm_new_listen_addr_total 0
# HELP libp2p_swarm_expired_listen_addr Number of expired listen addresses.
# TYPE libp2p_swarm_expired_listen_addr counter
libp2p_swarm_expired_listen_addr_total 0
# HELP libp2p_swarm_listener_closed Number of listeners closed.
# TYPE libp2p_swarm_listener_closed counter
libp2p_swarm_listener_closed_total 0
# HELP libp2p_swarm_listener_error Number of listener errors.
# TYPE libp2p_swarm_listener_error counter
libp2p_swarm_listener_error_total 0
# HELP libp2p_swarm_dial_attempt Number of dial attempts.
# TYPE libp2p_swarm_dial_attempt counter
libp2p_swarm_dial_attempt_total 0
# HELP libp2p_swarm_outgoing_connection_error Number outgoing connection errors.
# TYPE libp2p_swarm_outgoing_connection_error counter
# HELP libp2p_swarm_connected_to_banned_peer Number of connection attempts to banned peer.
# TYPE libp2p_swarm_connected_to_banned_peer counter
libp2p_swarm_connected_to_banned_peer_total 0
# HELP libp2p_swarm_connections_established Number of connections established.
# TYPE libp2p_swarm_connections_established counter
# HELP libp2p_swarm_connections_closed Number of connections closed.
# TYPE libp2p_swarm_connections_closed counter
# EOF
```

**Reference issue to close (if applicable)**

Closes https://github.com/ChainSafe/forest/issues/2253

**Other information and links**

### CODE REVIEW

**Review by lemmih — APPROVED** [review]

File: node/forest_libp2p/src/service.rs [hunk 1]
```diff
@@ -455,57 +460,60 @@ async fn handle_forest_behaviour_event<DB, P: StoreParams>(
...
+                        }
+                        Err(e) => {
+                            warn!(
+                                "Gossip Message from peer {:?} could not be deserialized: {}",
+                                source, e
+                            );
+                        }
                     }
+                } else {
+                    warn!("Getting gossip messages from unknown topic: {}", topic);
```
Comment (LesnyRumcajs) [inline node/forest_libp2p/src/service.rs:513]:
```suggestion
                    warn!("Getting gossip messages from unknown topic: {topic}");
```

File: node/forest_libp2p/src/service.rs [hunk 2]
```diff
@@ -455,57 +460,60 @@ async fn handle_forest_behaviour_event<DB, P: StoreParams>(
...
+                                    message: PubsubMessage::Block(b),
+                                },
+                            )
+                            .await;
+                        }
+                        Err(e) => {
+                            warn!(
+                                "Gossip Block from peer {:?} could not be deserialized: {}",
+                                source, e
+                            );
```
Comment (LesnyRumcajs) [inline node/forest_libp2p/src/service.rs:490]:
```suggestion
                            warn!(
                                "Gossip Block from peer {source:?} could not be deserialized: {e}",
                            );
```

File: node/forest_libp2p/src/service.rs [hunk 3]
```diff
@@ -455,57 +460,60 @@ async fn handle_forest_behaviour_event<DB, P: StoreParams>(
...
+                                    message: PubsubMessage::Message(m),
+                                },
+                            )
+                            .await;
+                        }
+                        Err(e) => {
+                            warn!(
+                                "Gossip Message from peer {:?} could not be deserialized: {}",
+                                source, e
+                            );
```
Comment (LesnyRumcajs) [inline node/forest_libp2p/src/service.rs:509]:
```suggestion
                            warn!(
                                "Gossip Message from peer {source:?} could not be deserialized: {e}");
```

File: utils/metrics/src/lib.rs [hunk 4]
```diff
@@ -26,15 +36,22 @@ pub async fn init_prometheus(
...
+    let metric_families = default_registry.gather();
     let mut metrics = vec![];
 
     let encoder = TextEncoder::new();
     encoder
         .encode(&metric_families, &mut metrics)
         .expect("Encoding Prometheus metrics must succeed.");
 
+    for (name, registry) in REGISTRIES_EXT.read().await.iter() {
+        metrics.extend_from_slice(format!("\n# SECTION {name}\n").as_bytes());
```
Comment (LesnyRumcajs) [inline utils/metrics/src/lib.rs:49]: Is this `SECTION` a standard Prometheus keyword? Does Grafana handle it? If not, I'd not add it.

**Review by LesnyRumcajs — APPROVED** [review]

File: utils/metrics/src/lib.rs (hunk 4 above)
Comment (hanabi1224) [reply utils/metrics/src/lib.rs:49]: I believe the leading `#` makes it a comment line, gGafana should just ignore it
https://prometheus.io/docs/prometheus/latest/querying/basics/#comments

Comment (LesnyRumcajs) [reply utils/metrics/src/lib.rs:49]: Thanks. Still, I wouldn't add custom stuff there. Grafana may be fine, I encountered parsers that were not happy with such things even though it was technically legal to add them. It also doesn't ignore it, e.g., those `# HELP` and `# TYPE` are used.

Comment (hanabi1224) [reply utils/metrics/src/lib.rs:49]: OK, let me remove it, blank lines should be fine?

Comment (LesnyRumcajs) [reply utils/metrics/src/lib.rs:49]: I would not add anything, vanilla metrics are the safest option.

Comment (LesnyRumcajs) [reply utils/metrics/src/lib.rs]: Thank you!

**Pull request merged**